h2/src/main/org/h2/util/DateTimeUtils.java

```diff
@@ -515,22 +515,13 @@ public static Value parseTimestamp(String s, Mode mode, boolean withTimeZone) {
             if (tz != null) {
                 if (withTimeZone) {
                     if (tz != UTC) {
-                        long millis = convertDateValueToMillis(UTC, dateValue);
+                        long millis = convertDateTimeValueToMillis(tz, dateValue, nanos / 1000000);
```

Comment:
TimeZone::getOffset takes a UTC date/time value, so this doesn't look right

Reply:
`convertDateTimeValueToMillis()` always returns milliseconds since UNIX epoch in UTC, not in specified time zone. This is internal Java representation for old APIs.

This method uses `tz` argument to read source date-time value.

Reply:
You can see results in a new test:
`2018-03-25 01:59:00 Europe/Berlin` -> `2018-03-25 01:59:00.0+01`
And one minute after (there is a transition from `2:00` to `3:00`):
`2018-03-25 03:00:00 Europe/Berlin` -> `2018-03-25 03:00:00.0+02`
Second timestamp takes summer time offset (+02), not winter one as previous. Before this change second timestamp was incorrectly parsed as `2018-03-25 03:00:00.0+01`.